pkg/containerutil/cp_resolve_linux.go

```diff
@@ -0,0 +1,444 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package containerutil
+
+import (
+	"errors"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"runtime"
+	"slices"
+	"strings"
+	"syscall"
+
+	"github.com/opencontainers/runtime-spec/specs-go"
+
+	"github.com/containerd/containerd/v2/pkg/oci"
+)
+
+// volumeNameLen returns length of the leading volume name on Windows.
+// It returns 0 elsewhere.
+// FIXME: whenever we will want to port cp to windows, we will need the windows implementation of volumeNameLen
+func volumeNameLen(_ string) int {
+	return 0
+}
+
+var (
+	errDoesNotExist           = errors.New("resource does not exist")                                            // when a path parent dir does not exist
+	errIsNotADir              = errors.New("is not a dir")                                                       // when a path is a file, ending with path separator
+	errCannotResolvePathNoCwd = errors.New("unable to resolve path against undefined current working directory") // relative host path, no cwd
+)
+
+// pathSpecifier represents a path to be used by cp
+// besides exposing relevant properties (endsWithSeparator, etc), it also provides a fully resolved *host* path to
+// access the resource
+type pathSpecifier struct {
+	originalPath         string
+	endsWithSeparator    bool
+	endsWithSeparatorDot bool
+	exists               bool
+	isADir               bool
+	readOnly             bool
+	resolvedPath         string
+}
+
+// getPathSpecFromHost builds a pathSpecifier from a host location
+// errors with errDoesNotExist, errIsNotADir, "EvalSymlinks: too many links", or other hard filesystem errors from lstat/stat
+func getPathSpecFromHost(originalPath string) (*pathSpecifier, error) {
+	pathSpec := &pathSpecifier{
+		originalPath:         originalPath,
+		endsWithSeparator:    strings.HasSuffix(originalPath, string(os.PathSeparator)),
+		endsWithSeparatorDot: filepath.Base(originalPath) == ".",
+	}
+
+	path := originalPath
+
+	// Path may still be relative at this point. If it is, figure out getwd.
+	if !filepath.IsAbs(path) {
+		cwd, err := os.Getwd()
+		if err != nil {
+			return nil, errors.Join(errCannotResolvePathNoCwd, err)
+		}
+		path = cwd + string(os.PathSeparator) + path
```

Comment:
`filepath.Join(cwd, path)`

Reply:
I do not fully remember the detailed reason of why in that case, but I am pretty sure it would not work. filepath.Join will clean the path, which we do not want in that case.

Reply:
Re-reading. We absolutely do not want filepath.Join here. That would defeat the whole purpose of resolving the final path properly.

Reply:
filepath.Join will "clean" the path (eg: reduce relative path). The whole problem and point here is to perform *proper* resolution instead.

Reply:
IMO, we should reflect this intention in a comment.

Reply:
> filepath.Join will "clean" the path (eg: reduce relative path). The whole problem and point here is to perform proper resolution instead.

Why `clean` will be an issue here? do you have an example/test that can show the issue?

Reply:
Because it will clean-up the path purely lexically with no regards for the reality of the files.

Something like `/doesnotexist/foo/../../etc/passwd` clearly illustrates the issue.

It will, and must fail when copying or reading from it, as proper path resolution will fail for any actual filesystem operation (cp, read, write), but filepath.Join will "cleanup" to something that exist.

Conversely, with:
```
mkdir -p /tmp/target1/target2
ln -s /tmp/target1/target2 tmp/symlink
touch /tmp/target1/somefile
```

The path `/tmp/symlink/../somefile` will fail to resolve to an existing file using `filepath.Join`, while it SHOULD, and will, using any filesystem operation, since it is actually a valid file.

It is a common misunderstanding about lexical functions - they should not be used before resolving symlinks.

Reply:
> It is a common misunderstanding about lexical functions - they should not be used before resolving symlinks.

isn't safer to resolve the symlink before and then work with cleaned path ?

Reply:
I see that you evaluate the symlink at this line https://github.com/containerd/nerdctl/pull/3323/files#diff-f8d3022dbe81c88660be3b5c26107695a00ef30d103f67558a218780e5da1dbaR81 so may be resolving the relative path just after it ?

Reply:
> > It is a common misunderstanding about lexical functions - they should not be used before resolving symlinks.
> 
> isn't safer to resolve the symlink before and then work with cleaned path ?

I am not completely sure what you mean - but here, the path may be relative, so in that case we do need to start resolving against cwd.

Reply:
> I see that you evaluate the symlink at this line https://github.com/containerd/nerdctl/pull/3323/files#diff-f8d3022dbe81c88660be3b5c26107695a00ef30d103f67558a218780e5da1dbaR81 so may be resolving the relative path just after it ?

We need to account for the fact that the path may not exist - which is handled a few lines below.
Again not completely sure what you mean here, but see previous comment ^ - we need to handle relative path cases (which is the reason we resolve links against cwd + path).

Reply:
Hi @apostasie, I will take another look when I get some time. 

> So, point is: I am not interested in forcing anything here.
> We all have better things to do.
> If you folks are not comfortable with this PR, I would just accept that and close it.

I don't think folks imply you're forcing anything or not comfortable with your change. I think mostly it's because people (including me) may be slow at responding recently due to holidays approaching.

Reply:
All fine @djdongjin 
I want(ed) to make sure that people feel comfortable with a “no”
I am a simple and straightforward person 😁

Enjoy the holiday folks!